README
MZ-Data-Extract is a simple tool that you can use for collect relevant data of Portable Executable (PE) files that can be used for Intel during a line of research related with malware. All information collected can be used for Intel purposes. It support PE 32bits and 64bits for executables files type EXE, DLL, SYS, SCR, CPL, MSI, COM and at the moment just work in *NIX-Based distros.

This tool should always be accompanied by the file "packerdb.txt" to work. If you try to get information from a file that is not MZ and/or don't use the mentioned file together with the tool, you will see the following message in red color: "It isn't a PE file or missing file packerdb.txt. Please, check it and try again."
Please check accordingly.

For historical information about this tool please read file "CHANGELOG". 
For samples tested information please read file "FILES_TESTED_TODO".
For know wath information you can obtain with this tools please read file "REPORT_EXAMPLE".

Usage syntaxis is: ./mzde.py [filepath]

* Requires:
First, remember run the following commands:
 apt-get update 
 apt-get upgrade
 apt-get install python-pip

Then just run "requirements.txt" file using the command pip install -r requirements.txt

But if you prefer to install the packages separately, follow the instructions for each case:

* PEfile: https://github.com/erocarrera/pefile 
 apt-get install python-pefile 

* Magic: https://pypi.python.org/pypi/python-magic
 pip install python-magic

* SSDeep: https://pypi.python.org/pypi/ssdeep/3.2
 pip install ssdeep

* FuzzyHashLib: https://pypi.python.org/pypi/fuzzyhashlib
 pip install fuzzyhashlib

This package is required for get SDHash data, but if you have any problem with package install process, can install SDHash separately:
    * SDHash: https://pypi.python.org/pypi/sdhash
      pip install sdhash
        
Maybe you need install requires distributions for SDHash separately. Please use following commands:
  pip install NumPy or visit https://pypi.python.org/pypi/numpy
  pip install SciPy or visit https://pypi.python.org/pypi/scipy
  pip install Pillow

***** By default, SDHash option is disabled because the string is too long. If you want to know this data, please uncomment the line 83, or line "print Fore.WHITE + Style.NORMAL + "SDHash: %s" % Style.DIM + fuzzyhashlib.sdhash(data).hexdigest()" and save change.

***** Fuzzy-Hash Algorithms just show data, this process don't compare files. 
 
* Bitstring: https://pypi.python.org/pypi/bitstring/3.1.5
 pip install bitstring

* argparser: https://code.google.com/archive/p/argparse/
 apt-get install python-argparse

* colorama: https://pypi.python.org/pypi/colorama/
 apt-get install python-colorama

DISCLAIMER: This tool was written for internal and personal use and tested, at the moment, just under Ubuntu 16.04. Please use the tool at your own risk.

AUTHOR: Jorge (Pistus) Mieres
Email: jamieres-[at]-gmail-[dot]-com. 
Twitter: @jorgemieres


CHANGELOG
2016/09/21 v0.1
=> Start code with support for EXE, SYS, DLL, SCR and CPL 32/64bits files.

2016/09/24 v0.2
=> Private release with following features: File Name, File Size, EntryPoint (hex), TimeStamp (hex), PE Signature (hex), Image Base (hex), Packer Information, PE Sections with following value Section Name, Virtual Address, Virtual Size & Raw Data, PE Imports.

2016/09/27 v0.3
=> Add following data: Machine, Dynamic-Link Library (DLL), Subsystem information, RVA Number/Size, TimeStamp human readable date, PE section number, Section/VirtualAddress/VirtualSize and RawData tag for PE sections data.

2016/09/30 v0.4
=> Add Pointer to Symbol Table, Number of Symbols, Size of Optional Header, Characteristics, Required CPU Type, Directories data.
=> Add EntryPoint and TimeStamp data in decimal format.

2016/10/01 v0.5
=> Add following features: Magic, Major/Minor Linker Version, Size of Code, Base of Code, Size Init/Uninit Data, Section Alignment, File Alignment, Major/Minor Operating System Version, Major/Minor Image Version, Major/Minor Subsystem Version, Size of Image and headers, Checksum, DLL Charactersitics, Size of Stack Reserve/Commit, Size of Heap Reserve/Commit and Loader Flags. 

2016/10/02 v0.6
=> Separate data structure with "PE File Basic Information", "PE File Header Information", "PE Sections Information" and "PE Import Address Table".

2016/10/03 v0.7
=> Add MD5, SHA1, SHA256 and SHA512 hashing data.
=> Add peHashNG based in work https://github.com/AnyMaster/pehashng.
=> Add TotalHash, a peHash implementation, based in Frank Poz's work https://github.com/knowmalware/pehash.
=> Add Export Address Table.

2016/10/04 v0.8
=> Add hexadecimal/decimal data for all values of "File Header Infomation".
=> PE structure organizaton for "Optional Header Information".
=> Add File Signature (aka "Magic Number").

2016/10/05 v0.9
=> Add File size value in megabytes. "Last Modified", "Last Accessed", "Creation Time data" and "Filepath" data.
=> Add "Analysis date" information in the first line of the output.

2016/10/06 v1.0
=> Add EntryPoint Adreess value in hexadecimal. 
=> Add report date.
=> Add EXE and SYS detect data.
=> Add Machine reference for hexadecimal value: [0x14c=I386] [0x0200=IA64] [0x8664=AMD64].
=> Add Metadata information (properties data).
=> Add SSDeep fuzzy-hashing algorithm data.
=> Add Filetype and Mimetype data (from Linux command-line).
=> Add Export Address Table data.
=> Fixed some problems with "IMAGE_FILE_MACHINE" output.
=> Add "VS_VERSIONINFO" and "VS_FIXEDFILEINFO" in Properties. 
=> Rewrited code structure
=> Rewrited "properties information" code.
=> Rewrited "Sections Information" code.
=> Rewrited "Import Address Table" code.
=> Add "Suspicious" data in "Sections Information" segment.
=> Add "Suspicious API Functions" data.
=> Add "URL Information" data.
=> Rewrited code structure.
=> Add Overlay Information.
=> Add "Embedded File Information" data segment.
=> Add MD5 hash information in each "section name" data of "Sections Information" code segment.
=> Add AntiDebugging information.
=> Fixed "Issue #1"!. Show the message 'PE' object has no attribute 'DIRECTORY_ENTRY_EXPORT' in file that no have export dll.
=> Delete "Suspicious API Functions". 
=> Rewrited code structure.

2016/10/30 v1.0.1
=> Add SSDeep fuzzy-hashing algorithm data.
=> Add Filetype and Mimetype data (from Linux command-line).
=> Add Export Address Table data.
=> Fixed some problems with "IMAGE_FILE_MACHINE" output.
=> Add "VS_VERSIONINFO" and "VS_FIXEDFILEINFO" in Properties. 
=> Rewrited code structure

2016/11/03 v1.0.2
=> Rewrited "properties information" code.
=> Rewrited "Sections Information" code.
=> Rewrited "Import Address Table" code.
=> Add "Suspicious" data in "Sections Information" segment.
=> Add "Suspicious API Functions" data.
=> Add "URL Information" data.
=> Rewrited code structure.


2016/11/06 v1.0.3
=> Add Overlay Information.
=> Add "Embedded File Information" data segment.
=> Add MD5 hash information in each "section name" data of "Sections Information" code segment.
=> Add AntiDebugging information.
=> Fixed "Issue show the message 'PE' object has no attribute 'DIRECTORY_ENTRY_EXPORT' in file that no have export dll.
=> Delete "Suspicious API Functions". 
=> Rewrited code structure.

2016/11/10 v1.0.4
=> Rewrited code structure.
=> Add "Evasion Techniques Information" section with "Anti-Debugging Techniques Founded", "Anti-VirtualMachine Tricks Founded", "DLL Hooking Strings Founded", "Kernel Driver Check Founded" and "Anti-AV Tricks Founded" evasion data.
=> Delete "Additional properties information" and "Technical Header Information" sections. Maybe I will add it again in future release!

2017/04/03 v1.0.5
=> Rewrited code structure.
=> Add ImportHash (via FireEye: https://www.fireeye.com/blog/threat-research/2014/01/tracking-malware-import-hashing.html)
=> Rewrited "Embedded File Information" code.

2017/04/07 v1.0.6
=> Delete PeHASH and TotalHASH information. Show two fuzzy-hash: SSDeep and ImportHASH.
=> Fixed problem message for different format files (no-PE files). Error: 'DOS Header magic not found.'
=> Add exception for files with different format (no-PE files).
=> Rewrited Evasion Techniques code for "Advanced Technical Information" section.
=> Add PDB pathways filename information.
=> Add signatures for Anti-Debbugger Techniques.
=> Add static registry information with detection of Autorun.inf files (USB propagation technique). 
=> Delete "DLL Hooking Strings Founded", "Kernel Driver Check Founded" and "Anti-AV Tricks Founded". 
=> Rewrite code structure.

2017/06/27 v1.0.7
=> Delete "Reference to Embedded File Information". I will rewrite regex data...
=> Rewrite "PDB pathways filename information" and change section name for "PDB full pathway data".
=> Rewrite code structure.

2017/07/07 v1.0.8

MZDE107.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

__description__ = 'MZ-Data-Extract is a simple tool that you can use for collect relevant data of Portable Executable (PE) files that can be used for Intel during a line of research related with malware.'
__version__ = '1.0.7'
__date__ = 'September 21, 2016 (first release [v0.1])'
__author__ = 'Jorge (Pistus) Mieres'
__contact__ = 'jamieres@gmail.com'

"""
Requires:
See file called README.

Changelog:
2016/11/06 v1.0.3
2016/11/10 v1.0.4
2017/04/03 v1.0.5

2017/04/07 v1.0.7
=> Delete "Reference to Embedded File Information".
=> Rewrite "PDB pathways filename information" and change section name for "PDB full pathway data".

** For complete reference about history of this tool please real file "CHANGELOG".
"""

import sys,os,re,subprocess,datetime,time,shlex,stat,colorama,hashlib,binascii,ssdeep,bz2,string,bitstring,struct,pefile,peutils,argparse
from colorama import Fore, Back, Style, init
from pefile import PE, PEFormatError
from time import gmtime, strftime
from hashlib import sha256
from bz2 import compress
from struct import pack
init()

try:
  filename = sys.argv[1]
  pe = pefile.PE(filename)
  packer = peutils.SignatureDatabase('packerdb.txt')

except:
        print(Fore.RED + "It isn't a PE file or missing file packerdb.txt. Please, check it and try again.")
        sys.exit(1)

else:
  filename = sys.argv[1]

with open(filename, "rb") as s:
      r = s.read()

over = pe.get_overlay_data_start_offset()
raw = pe.write()
data = open(sys.argv[1], "rb").read()
printable = set(string.printable)
file_stats = os.stat(filename)
ep = pe.OPTIONAL_HEADER.AddressOfEntryPoint
ep_ad = ep + pe.OPTIONAL_HEADER.ImageBase
ssd = ssdeep.hash(data)
magic = subprocess.check_output(["file", "-b", filename])
mimetype = subprocess.check_output(["file", "--mime-type", "-b", filename])
NETINT = open(filename, 'rb')
hosts= open(filename,'r').readlines()
adbg = ['AddVectoredExceptionHandler','CheckRemoteDebuggerPresent','CloseHandle','DbgBreakPoint','DbgUiRemoteBreakin','DebugActiveProcess','FindWindow','GenerateConsoleCtrlEvent','GetLocalTime','GetSystemTime','GetTickCount','GetWindowThreadProcessId','HideThreadFromDebugger','IsDebugged','IsDebuggerPresent','NtClose','NtGlobalFlag','NtQueryObject','NtQueryPerformanceCounter','NtQuerySystemInformation','NtSetInformationThread','ObsidianGUI','OutputDebugString','Process32First','Process32Next','QueryInformationProcess','QueryPerformanceCounter','ReadTEB','RemoveVectoredExceptionHandler','RockDebugger','SetConsoleCtrlHandler','SetInformationThread','SetThreadContext','SetUnhandledExceptionFilter','TerminateProcess','UnhandledExceptionFilter','UnhandledExceptionFilter','WinDbgFrameClass', 'ZwQueryInformationProcess','Zeta Debugger','ZwQueryInformation','ZwQuerySystemInformation','ZwSetInformationThread','timeGetTime','vbaExceptHandler','GetThreadContext','SetThreadContext','NtContinue','Trap flag','timeGetTime','rdtsc','ZwDebugActiveProcess']
avm = {"Red Pill":"\x0f\x01\x0d\x00\x00\x00\x00\xc3","VirtualPC trick":"\x0f\x3f\x07\x0b","VMCheck.dll":"\x45\xC7\x00\x01","VMCheck.dll for VirtualPC":"\x0f\x3f\x07\x0b\xc7\x45\xfc\xff\xff\xff\xff","Bochs & QEmu CPUID Trick":"\x44\x4d\x41\x63","Torpig VMM Trick": "\xE8\xED\xFF\xFF\xFF\x25\x00\x00\x00\xFF\x33\xC9\x3D\x00\x00\x00\x80\x0F\x95\xC1\x8B\xC1\xC3","Torpig (UPX) VMM Trick": "\x51\x51\x0F\x01\x27\x00\xC1\xFB\xB5\xD5\x35\x02\xE2\xC3\xD1\x66\x25\x32\xBD\x83\x7F\xB7\x4E\x3D\x06\x80\x0F\x95\xC1\x8B\xC1\xC3","Virtual Box":"VBox","VBoxVideo.sys":"VBoxVideo.sys","VIRTUALBOX":"VIRTUALBOX","Xen":"XenVMM","XenVMMXenVMM":"XenVMMXenVMM","xennet.sys":"xennet.sys","VMware":"WMvare","VMware trick":"VMXh","VMTools":"VMTools","VMMEMCTL":"VMMEMCTL","vmx_svga.sys":"vmx_svga.sys","vmmouse.sys":"vmmouse.sys","vmsrvc.exe":"vmsrvc.exe","vmtoolsd.exe":"vmtoolsd.exe","vmusrvc.exe":"vmusrvc.exe","vmwaretray.exe":"vmwaretray.exe","VmRemoteGuest.exe":"VmRemoteGuest.exe","QEMU":"QEMU","Ven_Red_Hat&Prod_VirtIO":"Ven_Red_Hat&Prod_VirtIO","DiskVBOX":"DiskVBOX","DiskVirtual":"DiskVirtual","Wine String Artifact (wine_get_unix_file_name)":"wine_get_unix_file_name","KVMKVMKVM":"KVMKVMKVM","prl hyperv":"prl hyperv","Microsoft Hv":"Microsoft Hv","HARDWARE\\Description\\System":"HARDWARE\\Description\\System","HARDWARE\\DEVICEMAP\\Scsi\\Scsi Port 0\\Scsi Bus 0\\Target Id 0\\Logical Unit Id 0":"HARDWARE\\DEVICEMAP\\Scsi\\Scsi Port 0\\Scsi Bus 0\\Target Id 0\\Logical Unit Id 0","SOFTWARE\\Oracle\\VirtualBox Guest Additions":"SOFTWARE\\Oracle\\VirtualBox Guest Additions","SOFTWARE\\VMware, Inc.\\VMware Tools":"SOFTWARE\\VMware, Inc.\\VMware Tools","\VMware\VMware Tools":"\VMware\VMware Tools","VirtualMachine":"VirtualMachine","vmicheartbeat":"vmicheartbeat","vmwareuser.exe":"vmwareuser.exe","vmwaretray.exe":"vmwaretray.exe","vmdebug":"vmdebug","vboxservice.exe":"vboxservice.exe","vboxtray.exe":"vboxtray.exe","VBoxMouse":"VBoxMouse","VBoxGuest":"VBoxGuest","VBoxSF":"VBoxSF",}
malreg = ["HKEY_CURRENT_USER","HKEY_CLASSES_ROOT","HKEY_LOCAL_MACHINE","autorun.inf","HKLM_SOFTWARE_MICROSOFT_WINDOWS_CURRENTVERSION_RUN"]
file_info = {"fsize": round((file_stats [stat.ST_SIZE] / 1024) /1024.0, 3),}

print Fore.BLUE + Style.BRIGHT + "\n### File Characterization"
print "-------------------------"
print Fore.WHITE + Style.NORMAL + "  Filename: " + Style.DIM + str(sys.argv[1])
print Fore.WHITE + Style.NORMAL + "  Filepath: " + Style.DIM + str(os.path.realpath(sys.argv[1]))
print Fore.WHITE + Style.NORMAL + "  Filesize: " + Style.DIM + str(os.path.getsize(sys.argv[1])) + " bytes", (Fore.GREEN + "(%(fsize)s Megabytes)" % file_info)
print Fore.WHITE + Style.NORMAL + "  Mimetype: " + Style.DIM + mimetype
print Fore.WHITE + Style.NORMAL + "  Filetype: " + Style.DIM + magic
print Fore.WHITE + Style.NORMAL + "  DLL file (Dynamic-Link Library): ", Style.DIM + str(pe.is_dll())
print Fore.WHITE + Style.NORMAL + "  EXE file (Executable): ", Style.DIM + str(pe.is_exe())
print Fore.WHITE + Style.NORMAL + "  SYS file (System): ", Style.DIM + str(pe.is_driver())
print Fore.BLUE + Style.BRIGHT + "\n  ## Fingerprint Information"
print Fore.WHITE + Style.NORMAL + "  MD5: " + Style.DIM + hashlib.md5(data).hexdigest()
print Fore.WHITE + Style.NORMAL + "  Sha1: " + Style.DIM + hashlib.sha1(data).hexdigest()
print Fore.WHITE + Style.NORMAL + "  Sha256: " + Style.DIM + hashlib.sha256(data).hexdigest()
print Fore.WHITE + Style.NORMAL + "  Sha512: " + Style.DIM + hashlib.sha512(data).hexdigest()
print Fore.BLUE + Style.BRIGHT + "\n    # Fuzzy-Hash Algorithm"
print Fore.WHITE + Style.NORMAL + "    SSDeep: " + Style.DIM + ssd
print Fore.WHITE + Style.NORMAL + "    ImpHASH: %s" % Style.DIM + pe.get_imphash()

print Fore.BLUE + Style.BRIGHT + "\n  ## Header & Packer Technique"
def parseheader(header):
  fields = struct.unpack("<HHHHHHHHHHHHHH", header)
  signature = "".join([chr(fields[0] & 0xFF), chr(fields[0] >> 8)])
  print Fore.WHITE + Style.NORMAL + "  Signature: %s" % Style.DIM + signature

if __name__ == "__main__": 
  parser = argparse.ArgumentParser()
  parser.add_argument("filename")
  args = parser.parse_args()
  filename = args.filename
  f = open(filename, "r")
  header = f.read(0x1C)
  parseheader(header)

packer = peutils.SignatureDatabase('packerdb.txt')
matches = packer.match_all(pe, ep_only = True)
print Fore.WHITE + Style.NORMAL + "  Packer/Compiler: " + Style.DIM + str(matches)

print Fore.BLUE + Style.BRIGHT + "\n\n### Advanced Information for Intel"
print "----------------------------------"
print Fore.WHITE + Style.NORMAL + "  TimeStamp:", Style.DIM + str(datetime.datetime.fromtimestamp(pe.FILE_HEADER.TimeDateStamp)) 
print Fore.WHITE + Style.NORMAL + "  EntryPoint: " + Style.DIM + hex(pe.OPTIONAL_HEADER.AddressOfEntryPoint)
print Fore.WHITE + Style.NORMAL + "  EntryPoint Address: " + Style.DIM + "0x%07x" % ep_ad

print Fore.BLUE + Style.BRIGHT + "\n  ## Properties Information"
if hasattr(pe, 'VS_VERSIONINFO'):
  if hasattr(pe, 'FileInfo'):
    for entry in pe.FileInfo:
      if hasattr(entry, 'StringTable'):
        for st_entry in entry.StringTable:
          for str_entry in st_entry.entries.items():
	     print Fore.WHITE + Style.NORMAL + "  " + str_entry[0] + ': ' + Style.DIM + str_entry[1]

print Fore.BLUE + Style.BRIGHT + "\n  ## Classic Autorun.inf file Data"
def autorun_info():
        for line in hosts:
                for calls in malreg:
                        if re.search(calls, line):
                                print Fore.WHITE + Style.NORMAL + "  " + calls,line
def acheck():
	autorun_info()
acheck()

print Fore.BLUE + Style.BRIGHT + "\n  ## URL Information"
def ab(stream):
    found_str = ""
    while True:
        data = stream.read(1024*4)
        if not data:
            break
        for char in data:
            if char in printable:
                found_str += char
            elif len(found_str) >= 4:
                yield found_str
                found_str = ""
            else:
                found_str = ""

def cd():

		for found_str in ab(NETINT):
			print found_str
		PEtoStr.close()

esomalint = 0
for found_str in ab(NETINT):
			url = re.findall("((http|ftp|mailto|telnet|ssh)(s){0,1}\:\/\/[\w|\/|\.|\#|\?|\&|\=|\-|\%]+)+", found_str, re.IGNORECASE | re.MULTILINE)
			if url:
				print Fore.WHITE + Style.NORMAL + "  " + url[0][0]
				esomalint = esomalint + 1

else:
	if esomalint == 0:
		print Fore.YELLOW + Style.DIM + "   ** No data in this section **"

print Fore.BLUE + Style.BRIGHT + "\n\n  ## PDB full pathway data"
NETINT = open(filename, 'rb')
esomalint2 = 0
for found_str in ab(NETINT):
   pistuspdb = re.findall(r"(\D*\w*\s*\S*\W*\.(pdb))+", found_str, re.IGNORECASE | re.MULTILINE | re.DOTALL | re.UNICODE | re.VERBOSE)

   if pistuspdb:
	print Fore.WHITE + Style.NORMAL + "  " + pistuspdb[0][0]
        esomalint2 = esomalint2 + 1

else:
	if esomalint2 == 0:
		print Fore.YELLOW + Style.DIM + "   ** No embeddedd files founded **"


print Fore.RED + Style.BRIGHT + "\n\n**** Check in this list the files shown in the URL section to discard false positives."
print Fore.BLUE + Style.BRIGHT + "\n\n### Detected possible Evasion-Techniques/Maneuvers commonly used by malware"
print "---------------------------------------------------------------------------"
print Fore.BLUE + Style.BRIGHT + "  ## Anti-Debugging Techniques Founded"
for dbg in pe.DIRECTORY_ENTRY_IMPORT:
		for imp in dbg.imports:
			if (imp.name != None) and (imp.name != ""):
				for anti in adbg:
					if imp.name.startswith(anti):
						print "  %s %s" % (Fore.RED + Style.NORMAL + hex(imp.address), imp.name)

print Fore.BLUE + Style.BRIGHT + "\n  ## Anti-VirtualMachine Techniques Founded"
with open(filename,"rb") as f:
    buf = f.read()
    for trick in avm:
        pos = buf.find(avm[trick])
        if pos > -1:
	     print ("  " + Fore.RED + Style.NORMAL + "0x%x %s") % (pos, trick)

print Fore.BLUE + Style.BRIGHT + "\n\n  ## Sections Information"
print Fore.BLUE + Style.NORMAL + "\t============================================================================================================"
print Fore.WHITE + "\tSection\t   VirtualAddress\tVirtualSize\tSizeofRawData\tCharacteristics\t  Suspicious"
print Fore.BLUE + "\t============================================================================================================"
for section in pe.sections:
			section.get_entropy()
			if section.SizeOfRawData == 0 or (section.get_entropy() > 0 and section.get_entropy() < 1) or section.get_entropy() > 9:
				suspicious = Fore.RED + Style.NORMAL + "Possible Malicious Action"
			else:
				suspicious = Fore.GREEN + Style.NORMAL + ""

			print Fore.WHITE + Style.NORMAL + "\t%s\t   %s\t\t%s\t\t%s\t\t%s\t  %s" % (section.Name, hex(section.VirtualAddress),hex(section.Misc_VirtualSize),section.SizeOfRawData,hex(section.Characteristics),suspicious)
			print "\tMD5:",section.get_hash_md5()
print Fore.BLUE + Style.NORMAL + "\t============================================================================================================\n"
print Fore.BLUE + Style.BRIGHT + "### Import/Export Address Table Information"
print "-------------------------------------------"
print Fore.BLUE + Style.BRIGHT + "\n  ## Import Address Table"
i = 1
for entry in pe.DIRECTORY_ENTRY_IMPORT:
                bool = 1  
                print Fore.BLUE + Style.BRIGHT + "     " + entry.dll
                for imp in entry.imports:
                        if bool:
                                print Fore.WHITE + Style.NORMAL + "\t", Fore.MAGENTA + hex(imp.address), Fore.WHITE + str(imp.name), 
                                bool = 0
                        else:
                                sys.stdout.write("%s%s%s%s" % ("\n\t", Fore.MAGENTA + hex(imp.address)," ", Fore.WHITE + str(imp.name))) 
                i += 1
		print

print Fore.BLUE + Style.BRIGHT + "\n\n  ## Export Address Table"
try:
  for exp in pe.DIRECTORY_ENTRY_EXPORT.symbols:
    print Fore.MAGENTA + Style.NORMAL + "\t", hex(pe.OPTIONAL_HEADER.ImageBase + exp.address), Fore.WHITE + Style.NORMAL + exp.name, exp.ordinal
except:
	print Fore.YELLOW + Style.DIM + "     ** No data in this section **"

print Fore.BLUE + Style.BRIGHT + "\n\n  ## Appended Data (Overlay)"
def CIC(expression):
    if callable(expression):
        return expression()
    else:
        return expression

def IFF(expression, valueTrue, valueFalse):
    if expression:
        return CIC(valueTrue)
    else:
        return CIC(valueFalse)

def NumberOfBytesHumanRepresentation(value):
    if value <= 1024:
        return '%s bytes' % value
    elif value < 1024 * 1024:
        return '%.1f KB' % (float(value) / 1024.0)
    elif value < 1024 * 1024 * 1024:
        return '%.1f MB' % (float(value) / 1024.0 / 1024.0)
    else:
        return '%.1f GB' % (float(value) / 1024.0 / 1024.0 / 1024.0)

if over == None:
        print Fore.YELLOW + Style.DIM + "     ** No data in this section ** " 

else:
        print Fore.BLUE + Style.BRIGHT + "   # Appended Data: " 
        print Fore.WHITE + Style.NORMAL + "  Offset: " + Style.DIM + "0x%08x" % over
        overlaySize = len(raw[over:])
        print Fore.WHITE + Style.NORMAL + "  Size: " + Style.DIM + "0x%08x [%s] %.2f%%" % (overlaySize, NumberOfBytesHumanRepresentation(overlaySize), float(overlaySize) / float(len(raw)) * 100.0)
        print Fore.WHITE + Style.NORMAL + "  MD5: " + Style.DIM + "%s" % hashlib.md5(raw[over:]).hexdigest()
	print Fore.WHITE + Style.NORMAL + "  SHA1: " + Style.DIM + "%s" % hashlib.sha1(raw[over:]).hexdigest()
        print Fore.WHITE + Style.NORMAL + "  SHA256: " + Style.DIM + "%s" % hashlib.sha256(raw[over:]).hexdigest()
        
        overlayMagic = raw[over:][:4]
        if type(overlayMagic[0]) == int:
            overlayMagic = "".join([chr(b) for b in overlayMagic])
        print Fore.WHITE + Style.NORMAL + "  Magic: " + Style.DIM + "%s %s" % (binascii.b2a_hex(overlayMagic), "".join([IFF(ord(b) >= 32, b, ".") for b in overlayMagic]))

        print Fore.BLUE + Style.BRIGHT + "\n   # Without Appended Data:"
        print Fore.WHITE + Style.NORMAL + "  MD5: " + Style.DIM + "%s" % hashlib.md5(raw[:over]).hexdigest()
        print Fore.WHITE + Style.NORMAL + "  SHA1: " + Style.DIM + "%s" % hashlib.sha1(raw[:over]).hexdigest()
        print Fore.WHITE + Style.NORMAL + "  SHA256: " + Style.DIM + "%s" % hashlib.sha256(raw[:over]).hexdigest()

print Fore.BLUE + Style.DIM + "\n###############################################################\t"
print Fore.BLACK + Style.BRIGHT + "\t  Malware" + Fore.BLUE + Style.NORMAL + "Intelligence" + Fore.WHITE + Style.NORMAL + "\twww.malwareint.com"
print Fore.YELLOW + Style.DIM + "   Analysis date: " + strftime("%A, %B %d, %Y [%I:%M:%S %p]\n", gmtime())

